### PR TITLE
Feature/add ti resources

### PR DESCRIPTION
**Pull request by boltandrke**
This commit will add dynamic TiKV and PD resource handling with TiDB cluster and closes issue #15  
All supported TiKV and PD configuration parameters are described [here](https://docs.pingcap.com/tidb/stable/dynamic-config)

### CODE REVIEW

**Comment by borissavelev** [comment]
@petoju sorry for pinging, could you please take a look?)

File: GNUmakefile
Comment (petoju) [inline GNUmakefile:57]: Could we do the same in Docker?

Your approach forces you to start considering arch, runs with `nohup`, depends on some mirror without chance to cache it, there may be potentially be issues with binary compatibility on newer GLIBC and so on.

File: mysql/resource_ti_config_variable.go
Comment (petoju) [inline mysql/resource_ti_config_variable.go:70]: What if `varValue` contains apostrophes? Even spitting out a warning is better than nothing.

File: mysql/resource_ti_config_variable_test.go
Comment (petoju) [inline mysql/resource_ti_config_variable_test.go:207]: Indent it please.

File: mysql/resource_ti_config_variable.go
Comment (petoju) [inline mysql/resource_ti_config_variable.go:170]: No newline after return, please.

File: mysql/resource_ti_config_defaults.go
Comment (petoju) [inline mysql/resource_ti_config_defaults.go:1]: Where is this file coming from? Could you document please, how to regenerate it?

File: mysql/resource_ti_config_variable.go
Comment (petoju) [inline mysql/resource_ti_config_variable.go:35]: Does changing a value need `ForceNew` and therefore destruction of the resource?

File: mysql/resource_ti_config_variable.go
Comment (petoju) [inline mysql/resource_ti_config_variable.go:40]: I believe changing `type` will most likely need `ForceNew` at least logically. But I don't insist on this.

File: mysql/resource_ti_config_variable_test.go
Comment (petoju) [inline mysql/resource_ti_config_variable_test.go:142]: I don't really understand - how is `DB` environment set?

File: mysql/resource_ti_config_variable_test.go
Comment (petoju) [inline mysql/resource_ti_config_variable_test.go:140]: Isn't `testAccGetConfigVar` good enough? I mean you can have a wrapper, but having almost the same code twice doesn't sound like a wise decision.

Can we just call the other function from inside here if nothing else?

File: mysql/resource_ti_config_variable_test.go
Comment (petoju) [inline mysql/resource_ti_config_variable_test.go:197]: Please indent.

File: mysql/resource_ti_config_variable.go
Comment (petoju) [inline mysql/resource_ti_config_variable.go:67]: This doesn't sound right. For DBs respecting data types, it depends on the type of key. Type of value may help (maybe that's your approximation), but it doesn't always work.

Doesn't TiDB respect it (like MySQL) in case you quote everything?

**Review by petoju — CHANGES_REQUESTED** [review]
@borissavelev sorry, I was on vacation for this week and I plan being on one in the next week.

Except for some details, it looks reasonably well. I can still merge it if I get reasonable replies or changes.

File: GNUmakefile
Comment (boltandrke) [reply GNUmakefile:57]: Yes, its possible with version 4.0 as described [here](https://docs.pingcap.com/tidb/v4.0/test-deployment-using-docker). But its not straight forward setup as tiup does. 
Starting from version 5.0 docker is not anymore supported method pulling up the cluster.  
This tool is needed only for acceptance testing. I'll glue it to am64 arch only. (similar like terraform binary fetch).

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go:70]: Here is an example If subsystem doesn't accept apostrophes

`mysql> set config pd `log.level`="`warn`";
Query OK, 0 rows affected, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                               |
+---------+------+---------------------------------------------------------------------------------------+
| Warning | 1105 | bad request to http://127.0.0.1:2379/pd/api/v1/config: "log level `warn` is illegal"
 |
+---------+------+---------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
`

It will be later rejected by warning check.

I'll add extra logging about this

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go:170]: 👍

File: mysql/resource_ti_config_defaults.go
Comment (boltandrke) [reply mysql/resource_ti_config_defaults.go:1]: Sure :)

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go:35]: True. I'll fix it!

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go:40]: yes, your are correct! I'll fix it :)

File: mysql/resource_ti_config_variable_test.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable_test.go:142]: With Travis tests Here https://github.com/petoju/terraform-provider-mysql/blob/master/.travis.yml#L11-L13

File: mysql/resource_ti_config_variable_test.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable_test.go:197]: 👍

File: mysql/resource_ti_config_variable_test.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable_test.go:140]: I'll make workaround. Only one test requires empty string for varInstance. All other tests requires real instance.

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go:67]: L64: 
If instance is in place then it requires extra quoting like in the example 
`
set config "127.0.0.1:20180"  `split.qps-threshold`=1000
`

Everything else should be configured like this
`set config tikv `split.qps-threshold`=1000`

ref. https://docs.pingcap.com/tidb/stable/dynamic-config

File: mysql/resource_ti_config_defaults.go
Comment (petoju) [reply mysql/resource_ti_config_defaults.go:10]: Thanks!

File: mysql/resource_ti_config_variable.go
Comment (petoju) [reply mysql/resource_ti_config_variable.go:40]: Resolved

File: mysql/resource_ti_config_variable.go
Comment (petoju) [reply mysql/resource_ti_config_variable.go:74]: Ok, my question was maybe laid incorrectly.

Can it happen, that some value will contain apostrophe = `'` character? In that case, your SQL will be invalid. BTW your new check for backticks `` ` `` seems to be broken (like it will warn almost every single time).

For example, is it possible or likely, that a human will want to set some value to `something'somethingelse`? If that is possible, then we should certainly add some escaping.

File: mysql/resource_ti_config_variable.go
Comment (petoju) [reply mysql/resource_ti_config_variable.go:67]: > L64: If instance is in place then it requires extra quoting like in the example ```set config "127.0.0.1:20180" `split.qps-threshold`=1000 ```
> 
> Everything else should be configured like this ```set config tikv `split.qps-threshold`=1000```

We don't understand each other - is it bad, if you quote it like this?
```
set config tikv `split.qps-threshold`='1000'
```
I believe it would be also acceptable and it would make it a bit easier to understand, but I'd leave it up to you.

File: mysql/resource_ti_config_variable_test.go
Comment (petoju) [reply mysql/resource_ti_config_variable_test.go:142]: Ah, ok then - but could you add a comment? Because that's completely opaque you want to skip travis runs.

File: GNUmakefile
Comment (petoju) [reply GNUmakefile:56]: Pingcap still publishes the docker-compose config: https://github.com/pingcap/tidb-docker-compose - but I admit it may be more complicated.

Still, if we don't want to use that - why should it run under `nohup`? Should it be left running in the background?

Comment (boltandrke) [reply GNUmakefile]: Replace tiup with minimal docker setup. PTAL

File: scripts/tidb-test-cluster.sh
Comment (petoju) [inline scripts/tidb-test-cluster.sh:101]: Consider also `os.Exit(1)` and checking for it when calling `go run`.

File: GNUmakefile
Comment (petoju) [reply GNUmakefile]: Thanks, this looks good. There is a minor issue there and I have / already had few more questions (replying you considered those is good enough for me). After that is done, I can merge it.

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go:78]: Thank you for pointing out this. I added validator for this.

File: mysql/resource_ti_config_variable.go
Comment (boltandrke) [reply mysql/resource_ti_config_variable.go]: Got it 😀 Thank you 👍 
Remove obsolete code

**Comment by boltandrke** [comment]
If I understood correctly, everything should be fixed now. 
Before you merge it, let me cleanup branch with rebase. :)

**Review by petoju — APPROVED** [review]
Looks good now - thanks.

I'm not merging it yet as you requested a rebase.

**Comment by boltandrke** [comment]
Thank you!🙏 Now its clean. All acceptance tests passed and we are 🟢 to merge it. 
Its was pleasure to work with you 🙇‍♂️

**Pull request merged**
